TFNetworkLayer.py

```diff
@@ -1044,16 +1046,33 @@ def batch_norm(self, data,
       https://github.com/deepmind/sonnet/blob/master/sonnet/python/modules/batch_norm.py
     """
     with tf.variable_scope("batch_norm"):
+      input_data = data.copy()
       if masked_time:
-        x = data.get_placeholder_flattened(keep_dims=True)
+        assert not use_frequency
+        # Default case
+        x = input_data.get_placeholder_flattened(keep_dims=True)
         mean, variance = tf.nn.moments(x, axes=[0], keep_dims=True)
+        variables_shape = input_data.get_bc_spatial_batch_shape()
       else:
-        x = data.placeholder
-        mean, variance = tf.nn.moments(x, axes=data.get_axes(exclude_feature=True), keep_dims=True)
+        # Perform batch_norm operations in NHWC format in order to
+        # keep consistency when saving/loading the variables.
+        if data.is_batch_feature_major:
+          # Transform the input to NHWC
+          input_data = data.copy_as_batch_spatial_major()
```

Comment:
I thought you want to have it efficient? This would make it quite inefficient.

Reply:
What would be your suggestions?

Reply:
It was efficient before? (But anyway, maybe also depends what the fused variant can handle.)

Reply:
I came up to 3 solutions:
1.  Perform all operations in one format (NHWC), and  transform the input to NHWC and transform the output back to the initial format (+ 2 transpose operations)
2.  Keep all variables in one format (NHWC) and transform them according to the input format (+ num_vars transpose operations, which is 4 by default)
3.  Keep all variables flat and reshape according to the input format (+num_vars reshape operations)
3d variant isn't good as it's hard to keep consistency with the old implementation
Now you say, that the 1st isn't efficient. But what solution would you suggest as efficient?

Reply:
Any wrong/incorrect solution is not a solution, so let's not discuss that (e.g. your flat variables).

It is not inefficient because I suddenly say so. You should see that for yourself. You also should be able to tell yourself which other solution is probably more efficient. Actually, the whole argument e.g. of using the fused op, or convolution on NCHW was for efficiency, so I wonder why you now think that efficiency is not important anymore?

Transforming the variables on-the-fly is also some overhead which you don't need. Although that is maybe not so relevant (much much faster than transforming the data here). Keep it in the format which is faster for training on GPU, then this does not give you any overhead for the training case.

I thought about detecting whether the parameters were saved in the other format and the automatically convert them (similarly as we automatically convert params in other cases, like CuDNN LSTM, or NativeLSTM/BlockLSTM, etc). But the on-the-fly transformation is probably simpler.

Maybe check first what the fused op really supports here, and just do it accordingly.

Reply:
Ping?